### PR TITLE
Added the ChainControl system

### PR DESCRIPTION
**Pull request by joshuahansel**
This is currently a duplication of the `ControlLogic` system from THM, which is currently co-existing rather than using the new system. I'm putting this PR up for a preliminary review, after which I will do all of the downstream removal/deprecation.

I currently call the new system "ChainControls", which derive from "Controls". They work with new, named data of type `ChainControlData<T>`. ChainControls can `declareChainControlData<T>("data_name")` and `getChainControlData<T>("data_name")`. The advantage to this system is that it allows the user to form chains/trees of control logic. For example, one ChainControl gets a sensor value from a Postprocessor, another gets another sensor value, a third computes some quantity with these as inputs, a forth performs a PID control on this computed quantity, a fifth adds a delay to the PID output signal, and finally a sixth uses that output signal to control a MOOSE object parameter.

System doc: https://mooseframework.inl.gov/docs/PRs/28964/site/syntax/ChainControls/index.html

Some TODOs:

- [x] Look into why THM forced `execute_on = INITIAL TIMESTEP_BEGIN`. Do we need this?
- [x] Add the dependency ordering
- [x] Add the integrity checking (were all retrieved data actually declared somewhere?)

Refs #28948

### CODE REVIEW

**Comment by loganharbour** [comment]
Can you share a high level summary of why this is needed? I'm sorry, I'm a bit naive here and haven't actually seen this in practice and it's hard to follow without a more physical example.

**Comment by joshuahansel** [comment]
@loganharbour I modified the description to give a high-level overview and example.

File: framework/include/actions/AddChainControlAction.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "AddControlAction.h"
+
+/**
+ * Adds a ChainControl.
```
[the remark this answers is not in the document]
Comment (joshuahansel) [reply framework/include/actions/AddChainControlAction.h:15]: I'll do "to the control warehouse". The ChainControlSystem doesn't come into play until those ChainControls start adding their data. It might be better named "ChainControlDataSystem" if people think that's appropriate.

[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/actions/AddChainControlAction.h:15]: Makes sense

File: test/tests/chaincontrols/get_postprocessor_chain_control/gold/get_postprocessor_chain_control_out.csv [hunk 2]
```diff
@@ -0,0 +1,5 @@
+time,function_pp,test_pp
+0,5,0
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply test/tests/chaincontrols/get_postprocessor_chain_control/gold/get_postprocessor_chain_control_out.csv:2]: If it s UOs and then controls I think this PP ought to have a warning in the doco as it will always be lagged?

**Comment by joshuahansel** [comment]
@GiudGiud @lindsayad @loganharbour This one is ready for review. Any takers?

File: framework/doc/content/source/actions/AddChainControlAction.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# AddChainControlAction
+
+This action adds a [ChainControl](syntax/ChainControls/index.md) to the the `ControlWarehouse`.
```
Comment (GiudGiud) [inline framework/doc/content/source/actions/AddChainControlAction.md:3]:
```suggestion
This action adds a [ChainControl](syntax/ChainControls/index.md) to the `ControlWarehouse`.
```

**Review by GiudGiud — COMMENTED** [review]
there's a couple files I ll take a second look at later

**Comment by joshuahansel** [comment]
OpenMPI likes to fail

File: framework/include/base/ChainControlDataSystem.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Queries if the chain control data of the given name exists.
+   *
+   * @param[in] data_name  Chain control data name
+   * @return True if the chain control data of the given name exists
+   */
+  bool hasChainControlData(const std::string & data_name) const
+  {
+    return _chain_control_data_map.find(data_name) != _chain_control_data_map.end();
```
Comment (lindsayad) [inline framework/include/base/ChainControlDataSystem.h:36]: I would put all these methods as inline methods outside of the class definition. Want perusal of the class definition to be quick

File: framework/include/base/ChainControlDataSystem.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                   "' was requested with the type '",
+                   MooseUtils::prettyCppType<T>(),
+                   "' but has the type '",
+                   chain_control_data_value->type(),
+                   "'.");
+      }
+    }
+    else
+    {
+      data = new ChainControlData<T>(_app, data_name);
```
Comment (lindsayad) [inline framework/include/base/ChainControlDataSystem.h:86]: What is this `new` business. Is this 2010? 😄

File: framework/include/chaincontrols/ChainControl.h [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  /// List of chain control data that this control depends upon
+  std::vector<std::string> _control_data_depends_on;
+};
+
+template <typename T>
+T &
+ChainControl::declareChainControlData(const std::string & data_name, bool apply_object_prefix)
+{
+  std::string full_data_name = (apply_object_prefix ? name() + ":" : "") + data_name;
```
Comment (lindsayad) [inline framework/include/chaincontrols/ChainControl.h:108]:
```suggestion
  const std::string full_data_name = (apply_object_prefix ? name() + ":" : "") + data_name;
```

File: framework/include/chaincontrols/ChainControlData.h [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Abstract definition of a ChainControlData value.
+ */
+class ChainControlDataBase : public Restartable
+{
+public:
+  /**
+   * Constructor
+   * @param moose_app MOOSEApp object this object belong to
```
Comment (lindsayad) [inline framework/include/chaincontrols/ChainControlData.h:24]:
```suggestion
   * @param moose_app MooseApp object this object belong to
```

**Comment by joshuahansel** [comment]
Ready for another round on this one.

File: framework/src/postprocessors/ChainControlDataPostprocessor.C [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+Real
+ChainControlDataPostprocessor::getValue() const
+{
+  if (_real_data)
+    return _real_data->get();
+  else if (_bool_data)
+    // for booleans, 'true' should convert to 1.0 and 'false' to 0.0.
+    return _bool_data->get();
+  else
+    mooseError("This point should not be reachable.");
```
Comment (GiudGiud) [inline framework/src/postprocessors/ChainControlDataPostprocessor.C:72]:
```suggestion
    mooseAssert(false, "This point should not be reachable.");
```

**Review by GiudGiud — COMMENTED** [review]
looks good to me. 

The only question is whether this is "frameworkey" or misc or could be part of a new module, and for now I vote framework.

File: framework/include/chaincontrols/ChainControl.h [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+   */
+  void addChainControlDataDependency(const std::string & data_name);
+
+  /// List of chain control data that this control depends upon
+  std::vector<std::string> _control_data_depends_on;
+};
+
+template <typename T>
+T &
+ChainControl::declareChainControlData(const std::string & data_name, bool apply_object_prefix)
```
[the remark this answers is not in the document]
Comment (joshuahansel) [reply framework/include/chaincontrols/ChainControl.h:106]: THM has this action `PostprocessorAsControlAction` that automatically adds a control data for each PP that gets added (but only those in the input file 🤮 ), and in THM the control `CopyPostprocessorValueControl` always named  the data to be the same as the PP. In the chain control `GetPostprocessorChainControl` I give the option to name it one way or the other. The idea was backwards compatibility with THM.

However, I don't think I personally like `PostprocessorAsControlAction`, since I'd rather be explicit about introducing control data into the system, even though it can be less convenient, so I'm open to trying to get rid of it in THM.

That said, I don't know if we should force the naming convention on applications, even though we think it wise (and having the prefix is the default at least).

**Comment by joshuahansel** [comment]
> The only question is whether this is "frameworkey" or misc or could be part of a new module, and for now I vote framework.

I think it should go in the framework - just a new branch of Controls.

File: framework/src/postprocessors/ChainControlDataPostprocessor.C (hunk 8 above)
Comment (joshuahansel) [reply framework/src/postprocessors/ChainControlDataPostprocessor.C:72]: Why don't we want an error in opt?

File: framework/include/chaincontrols/ChainControl.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/chaincontrols/ChainControl.h:106]: >  backwards compatibility with THM.

that is a good idea. If the chainControl does not get adopted by THM and the apps, we've created some debt by duplicating the system

> However, I don't think I personally like PostprocessorAsControlAction

me neither. Adding functor support to the ParsedControl could help you support PPs without a ChainControl intermediate object ?

File: framework/src/postprocessors/ChainControlDataPostprocessor.C (hunk 8 above)
Comment (GiudGiud) [reply framework/src/postprocessors/ChainControlDataPostprocessor.C:72]: this will never be hit since it's caught earlier  when initializing the data
and the error in opt will cost you 1 line of coverage

File: framework/include/chaincontrols/ChainControl.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshuahansel) [reply framework/include/chaincontrols/ChainControl.h:106]: Yeah `ParsedControl` should be able to be extended to include PPs. I don't know yet if that would be using functors or not. I get confused for contexts that are not on an element/side/etc.

For now, `GetPostprocessorChainControl` gets the job done.

**Comment by joshuahansel** [comment]
@lindsayad Comments should be addressed

File: framework/include/base/ChainControlDataSystem.h [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+   * data is now set to be declared, and the declaring ChainControl is captured.
+   *
+   * @tparam T                  Type of the data
+   * @param[in] data_name       Chain control data name
+   * @param[in] chain_control   Chain control that is declaring the data
+   * @return Reference to the declared chain control data
+   */
+  template <typename T>
+  ChainControlData<T> & declareChainControlData(const std::string & data_name,
+                                                ChainControl * chain_control);
```
Comment (lindsayad) [inline framework/include/base/ChainControlDataSystem.h:71]: why is this a pointer? Is it allowed to be null?

File: framework/include/functions/ChainControlParsedFunctionWrapper.h [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  const std::vector<ChainControlData<Real> *> getRealChainControlData()
+  {
+    return _real_control_data_values;
+  }
+
+  /**
+   * Get list of boolean-valued control data objects
+   */
+  const std::vector<ChainControlData<bool> *> getBoolChainControlData()
```
Comment (lindsayad) [inline framework/include/functions/ChainControlParsedFunctionWrapper.h:65]: Let's not return by `const` value

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/include/base/ChainControlDataSystem.h (hunk 10 above)
Comment (joshuahansel) [reply framework/include/base/ChainControlDataSystem.h:71]: Doesn't seem to be any good reason - I'll change it

**Review by lindsayad — APPROVED** [review]

**Review by GiudGiud — APPROVED** [review]
